### PR TITLE
feat(treesitter): improved logging

### PR DESCRIPTION
**Pull request by lewis6991**
- Add bindings to Treesitter ts_parser_set_logger and ts_parser_logger
- Add logfile with path STDPATH('log')/treesitter.c
- Rework existing LanguageTree loggin to use logfile
- Begin implementing log levels for vim.g.__ts_debug

### CODE REVIEW

File: runtime/lua/vim/fs.lua [hunk 1]
```diff
@@ -73,10 +73,14 @@ function M.basename(file)
 end
 
 ---@private
-local function join_paths(...)
+---@param ... string
+---@return string
+function M._join_paths(...)
```
Comment (justinmk) [inline runtime/lua/vim/fs.lua:78]: Any reason we haven't exposed this as `fs.join()` ?

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 2]
```diff
@@ -31,8 +31,17 @@
...
+---
+
+-- Debugging:
+--
+-- vim.g.__ts_debug levels:
+--    - 1. Messages from languagetree.lua
+--    - 2. Parse messages from treesitter
+--    - 2. Lex messages from treesitter
+--
+-- Log file can be found in STDPATH('log')/treesitter.log
```
Comment (justinmk) [inline runtime/lua/vim/treesitter/languagetree.lua:43]: why is STDPATH uppercase? :)
```suggestion
-- Log file can be found in stdpath('log')/treesitter.log
```

File: src/nvim/lua/treesitter.c [hunk 3]
```diff
@@ -669,9 +677,85 @@ static int parser_get_timeout(lua_State *L)
...
+    return;
+  }
+
+  MAXSIZE_TEMP_ARRAY(args, 2);
+  ADD_C(args, STRING_OBJ(cstr_as_string(logtype == TSLogTypeParse ? "parse" : "lex")));
+  ADD_C(args, STRING_OBJ(cstr_as_string((char *)s)));
+  nlua_call_ref_with_lstate(opts->lstate, opts->cb, NULL, args, false, NULL);
+}
+
+static int parser_set_logger(lua_State *L)
```
Comment (justinmk) [inline src/nvim/lua/treesitter.c:704]: what is the high-level docstring for this? Is the logger per-parser?

should this be quasi-private? Once we have a logging API maybe we won't need this?

Comment (lewis6991) [reply src/nvim/lua/treesitter.c:704]: >  Is the logger per-parser?

Yes, this is implied by the function prefix and the fact it is contained in `parser_meta`.

> should this be quasi-private? Once we have a logging API maybe we won't need this?

Yeah maybe. This file is just dumb bindings into treesitter and the *real* API is managed in Lua. I'll prefix it with an underscore.

File: runtime/lua/vim/fs.lua (hunk 1 above)
Comment (clason) [reply runtime/lua/vim/fs.lua:78]: (nvim-treesitter also implements path handling; would be great to replace that with `vim.fs` as much as possible)

**Pull request merged**